src/main/java/io/temporal/internal/statemachines/ChildWorkflowStateMachine.java

```diff
@@ -0,0 +1,289 @@
+   *     ChildWorkflowExecutionTerminatedEvent.
+   * @return cancellation callback that should be invoked to cancel the child
+   */
+  public static ChildWorkflowStateMachine newInstance(
+      StartChildWorkflowExecutionCommandAttributes attributes,
+      Functions.Proc1<WorkflowExecution> startedCallback,
+      Functions.Proc2<Optional<Payloads>, Exception> completionCallback,
+      Functions.Proc1<NewCommand> commandSink) {
+    return new ChildWorkflowStateMachine(
+        attributes, startedCallback, completionCallback, commandSink);
+  }
+
+  private ChildWorkflowStateMachine(
+      StartChildWorkflowExecutionCommandAttributes startAttributes,
+      Functions.Proc1<WorkflowExecution> startedCallback,
+      Functions.Proc2<Optional<Payloads>, Exception> completionCallback,
+      Functions.Proc1<NewCommand> commandSink) {
+    super(newStateMachine(), commandSink);
+    this.startAttributes = startAttributes;
+    this.startedCallback = startedCallback;
+    this.completionCallback = completionCallback;
+    action(Action.SCHEDULE);
+  }
+
+  public void createStartChildCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION)
+            .setStartChildWorkflowExecutionCommandAttributes(startAttributes)
+            .build());
+  }
+
+  public boolean isCancellable() {
+    return State.START_COMMAND_CREATED == getState();
+  }
+
+  /**
+   * Cancellation through this class is valid only when start child workflow command is not sent
+   * yet. Cancellation of an initiated child workflow is done through CancelExternalCommands. So all
+   * of the types besides ABANDON are treated differently.
+   */
+  public void cancel() {
+    action(Action.CANCEL);
+  }
+
+  private void cancelStartChildCommand() {
+    cancelInitialCommand();
+    CanceledFailure failure = new CanceledFailure("Child canceled", null, null);
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyCompleted() {
+    ChildWorkflowExecutionCompletedEventAttributes attributes =
+        currentEvent.getChildWorkflowExecutionCompletedEventAttributes();
+    Optional<Payloads> result =
+        attributes.hasResult() ? Optional.of(attributes.getResult()) : Optional.empty();
+    completionCallback.apply(result, null);
+  }
+
+  private void notifyStartFailed() {
+    StartChildWorkflowExecutionFailedEventAttributes attributes =
+        currentEvent.getStartChildWorkflowExecutionFailedEventAttributes();
+    Exception failure =
+        new ChildWorkflowTaskFailedException(
+            currentEvent.getEventId(),
+            WorkflowExecution.newBuilder().setWorkflowId(attributes.getWorkflowId()).build(),
+            attributes.getWorkflowType(),
+            RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+            null);
+    failure.initCause(
+        new WorkflowExecutionAlreadyStarted(
```

Comment:
is it guaranteed that this is the only reason why we could not start a workflow?

Reply:
Yes, according to [StartChildWorkflowExecutionFailedCause](https://github.com/temporalio/api/blob/4fd842b5ee4275d63fde0849c3b06803a1754131/temporal/api/enums/v1/failed_cause.proto#L60:6)